src/jit/lowerarm64.cpp

```diff
@@ -1018,11 +786,352 @@ void Lowering::TreeNodeInfoInit(GenTree* stmt)
             info->internalIntCount = 1;
             break;
         } // end switch (tree->OperGet())
-    
-        tree = next;
 
         // We need to be sure that we've set info->srcCount and info->dstCount appropriately
-        assert(info->dstCount < 2);
+        assert((info->dstCount < 2) || tree->IsMultiRegCall());
+
+        tree = next;
+    }
+}
+//------------------------------------------------------------------------
+// TreeNodeInfoInitReturn: Set the NodeInfo for a GT_RETURN.
+//
+// Arguments:
+//    tree      - The node of interest
+//
+// Return Value:
+//    None.
+//
+void
+Lowering::TreeNodeInfoInitReturn(GenTree* tree)
+{
+    TreeNodeInfo* info = &(tree->gtLsraInfo);
+    LinearScan* l = m_lsra;
+    Compiler* compiler = comp;
+
+    GenTree* op1 = tree->gtGetOp1();
+    regMaskTP useCandidates = RBM_NONE;
+
+    info->srcCount = (tree->TypeGet() == TYP_VOID) ? 0 : 1;
+    info->dstCount = 0;
+
+    if (varTypeIsStruct(tree))
+    {
+        // op1 has to be either an lclvar or a multi-reg returning call
+        if ((op1->OperGet() == GT_LCL_VAR) || (op1->OperGet() == GT_LCL_FLD))
+        {
+            GenTreeLclVarCommon* lclVarCommon = op1->AsLclVarCommon();
+            LclVarDsc* varDsc = &(compiler->lvaTable[lclVarCommon->gtLclNum]);
+            assert(varDsc->lvIsMultiRegRet);
+
+            // Mark var as contained if not enregistrable.
+            if (!varTypeIsEnregisterableStruct(op1))
+            {
+                MakeSrcContained(tree, op1);
+            }
+        }
+        else
+        {
+            noway_assert(op1->IsMultiRegCall());
+
+            ReturnTypeDesc* retTypeDesc = op1->AsCall()->GetReturnTypeDesc();
+            info->srcCount = retTypeDesc->GetReturnRegCount();
+            useCandidates = retTypeDesc->GetABIReturnRegs();
+        }
```

Comment:
The method was copied from the version in lowerxarch.cpp which has the same issue.

I believe that it is correct to move this statement out so that it always executes:

        if (useCandidates != RBM_NONE)
        {
            tree->gtOp.gtOp1->gtLsraInfo.setSrcCandidates(l, useCandidates);
        }